src/api/java/io/github/cvc5/api/Solver.java

```diff
@@ -31,14 +31,32 @@ public long getPointer()
 
   public void deletePointer()
   {
-    deletePointer(pointer);
+    if (pointer != 0)
+    {
+      deletePointer(pointer);
+    }
+    pointer = 0;
   }
 
   private static native void deletePointer(long pointer);
 
-  @Override public void finalize()
+  // store pointers for terms, sorts, etc
+  List<AbstractPointer> abstractPointers = new ArrayList<>();
+
+  @Override public void close()
+  {
+    // delete heap memory for terms, sorts, etc
+    for (AbstractPointer abstractPointer : abstractPointers)
+    {
+      abstractPointer.deletePointer();
+    }
```

Comment:
It struck me that this loop makes us free objects in the same order as they were created, while one would usually use the opposite order. We might be able to argue that the objects have no inter-dependencies and thus the destruction order does not matter.
I'd suggest to simply delete in reverse order, this should always be safe, even if we introduce inter-dependencies.

Reply:
Done.